### PR TITLE
Enable h-loop-fusion when control flow in the graph. 

### PR DESCRIPTION
**Pull request by nouiz**
This is a follow up from https://github.com/tensorflow/tensorflow/pull/57087
When using multi-GPU, control flow end up being added in the graph. So the optimization wasn't used in that case.
This PR make that optimization work when there is control flow in the graph.
But it ignore nodes that hare any control flow marker (being the predecessor or the successor).

@cheshire @akuegel

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/gpu_compiler.cc [hunk 1]
```diff
@@ -738,8 +738,6 @@ Status GpuCompiler::PrepareHloModuleForIrEmitting(HloModule* hlo_module) {
   pipeline.AddPass<CopyInsertion>(GetCanShareBuffer());
   // To fuse the copy.
   pipeline.AddPass<GpuHorizontalLoopFusion>("copy_");
-  // To remove temporary fused_computation created by GpuHorizontalLoopFusion
-  pipeline.AddPass<HloDCE>();
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:742]: What's the motivation behind removing that pass?

**Review by cheshire — APPROVED** [review]

File: tensorflow/compiler/xla/service/gpu/gpu_compiler.cc (hunk 1 above)
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:742]: It was discussed here: https://github.com/tensorflow/tensorflow/pull/57087#issuecomment-1276240281

Mostly, it was added as h-loop-fusion was not cleaning things up. But I and Adrian modified it to generates a clean graph. So it isn't needed anymore.

Comment (akuegel) [reply tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:742]: Agree, this is not needed anymore :)

File: tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion_test.cc [hunk 2]
```diff
@@ -650,6 +650,49 @@ ENTRY e {
   EXPECT_TRUE(RunAndCompare(hlo_text, std::nullopt));
 }
 
+  TEST_F(HorizontalLoopFusionTest, CopyInsertionFusionControlFlow) {
+  if (GetTestPlatform()->Name() != "CUDA") {
```
[the remark this answers is not in the document]
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/horizontal_loop_fusion_test.cc]: Done.

**Review by akuegel — APPROVED** [review]

**Pull request merged**